pkg/controller/control.go

```diff
@@ -1244,31 +1244,99 @@ func (c *Controller) handleErrorNoLock(err error) error {
 	return err
 }
 
-func (c *Controller) handleDiskNoSpaceErrorForReplicas(replicaNoSpaceErrMap map[string]string) error {
+func (c *Controller) handleDiskNoSpaceErrorForReplicas(replicaNoSpaceErrMap map[string]int) error {
 	if len(replicaNoSpaceErrMap) == 0 {
 		return nil
 	}
 
-	replicaWritableCount := 0
-	for _, r := range c.replicas {
-		if r.Mode == types.ERR {
-			continue
-		}
-		replicaWritableCount++
-	}
+	replicaRWCount, rwReplicasOnEnospcMap := c.getRWReplicaCountAndNoSpaceMap(replicaNoSpaceErrMap)
 
 	// https://github.com/longhorn/longhorn/issues/10718#issuecomment-3190666485
 	// If a replica is on a disk that is out of space, and others are not, this replica will be marked as ERR to avoid the inconsistent data.
-	// The volume will be readable when all disks of the replicas are out of space so last one replica will not be marked as ERR because of ENOSPC.
-	if len(replicaNoSpaceErrMap) == replicaWritableCount {
+	// Last one replica will not be marked as ERR for an ENOSPC error so the volume will still be readable when all disks of the replicas are out of space.
+	// If all replicas are on disks that are out of space at the same time, keep a list of replicas that have the same write bytes RW and other replicas will be set to ERR.
+	if len(rwReplicasOnEnospcMap) == replicaRWCount {
+		if replicaRWCount != 1 {
+			// the ErrNoSpaceLeftOnDevice error will be translated to TypeENOSPC/ENOSPC and handled by `liblonghorn`
+			rwReplicasToErrOnEnospcList := filterReplicasToErrOnEnospc(replicaNoSpaceErrMap)
+			for _, address := range rwReplicasToErrOnEnospcList {
+				c.setReplicaModeNoLock(address, types.ERR)
+			}
+		}
 		return types.ErrNoSpaceLeftOnDevice
 	}
-	for address := range replicaNoSpaceErrMap {
+
+	for address := range rwReplicasOnEnospcMap {
 		c.setReplicaModeNoLock(address, types.ERR)
 	}
+
 	return nil
 }
 
+func (c *Controller) getRWReplicaCountAndNoSpaceMap(replicaNoSpaceMap map[string]int) (int, map[string]int) {
+	replicaRWCount := 0
+	noSpaceRWReplicaMap := make(map[string]int, len(c.replicas))
+
+	for _, r := range c.replicas {
+		if r.Mode == types.RW {
+			replicaRWCount++
+			if wbn, exist := replicaNoSpaceMap[r.Address]; exist {
+				noSpaceRWReplicaMap[r.Address] = wbn
+			}
+			continue
+		}
+		if r.Mode == types.WO {
+			if _, exist := replicaNoSpaceMap[r.Address]; exist {
+				c.setReplicaModeNoLock(r.Address, types.ERR)
+			}
+			continue
+		}
+	}
+
+	return replicaRWCount, noSpaceRWReplicaMap
+}
+
+func filterReplicasToErrOnEnospc(replicaToErrOnNoSpaceMap map[string]int) []string {
+	wbReplicaMap := make(map[int]map[string]bool, len(replicaToErrOnNoSpaceMap))
+	for address, wb := range replicaToErrOnNoSpaceMap {
+		if wbReplicaMap[wb] == nil {
+			wbReplicaMap[wb] = make(map[string]bool)
+		}
+		wbReplicaMap[wb][address] = true
+	}
+
+	maxLenSetReplica := 0
```

Comment:
What do `maxLenSetReplica` and `maxLenWbList` mean?

Reply:
- `maxLenSetReplica`: the max number of replicas for a written byte number.
For example, wbReplicasMap: {1: {"a", "b"}, 2: {"c"}, 3: {"e", "f"}}
`maxLenSetReplica` will be 2.
- `maxLenWbList`: as the example above, the list will be `{1, 3}`
because 1 and 3 have the same `maxLenSetReplica`.